ntp-udp/src/raw_socket.rs

```diff
@@ -190,15 +204,15 @@ mod recv_message {
         let mut buf_slice = IoSliceMut::new(packet_buf);
         let mut addr = zeroed_sockaddr_storage();
 
-        let mut mhdr = libc::msghdr {
-            msg_control: control_buf.as_mut_ptr().cast::<libc::c_void>(),
-            msg_controllen: control_buf.len() as _,
-            msg_iov: (&mut buf_slice as *mut IoSliceMut).cast::<libc::iovec>(),
-            msg_iovlen: 1,
-            msg_flags: 0,
-            msg_name: (&mut addr as *mut libc::sockaddr_storage).cast::<libc::c_void>(),
-            msg_namelen: std::mem::size_of::<libc::sockaddr_storage>() as u32,
-        };
+        let mut mhdr = empty_msghdr();
+
+        mhdr.msg_control = control_buf.as_mut_ptr().cast::<libc::c_void>();
+        mhdr.msg_controllen = control_buf.len() as _;
+        mhdr.msg_iov = (&mut buf_slice as *mut IoSliceMut).cast::<libc::iovec>();
+        mhdr.msg_iovlen = 1;
+        mhdr.msg_flags = 0;
+        mhdr.msg_name = (&mut addr as *mut libc::sockaddr_storage).cast::<libc::c_void>();
+        mhdr.msg_namelen = std::mem::size_of::<libc::sockaddr_storage>() as u32;
```

Comment:
I wrote it in the more verbose style because the current style would enable undefined behavior. This creates unaligned references. See https://doc.rust-lang.org/stable/std/ptr/macro.addr_of_mut.html

See also https://lucumr.pocoo.org/2022/1/30/unsafe-rust/

Reply:
I think this is fine, because all of the fields are `repr(C)` and `0` is a valid value. Therefore we can safely use `MaybeUninit::zeroed().assume_init()`. Now we have a valid value of type `msghdr` on the stack, and setting fields on should be fine as well. I have verified this with miri. 

I believe that the example given does not work because `String` cannot be initialized with all zeros. The internal implementation uses this definition:

```rust
pub(crate) struct RawVec<T, A: Allocator = Global> {
    ptr: Unique<T>,
    cap: usize,
    alloc: A,
}
```

and a null pointer is not a valid value for `Unique<T>`, because

```
pub struct Unique<T: ?Sized> {
    pointer: NonNull<T>,
    // NOTE: this marker has no consequences for variance, but is necessary
    // for dropck to understand that we logically own a `T`.
    //
    // For details, see:
    // https://github.com/rust-lang/rfcs/blob/master/text/0769-sound-generic-drop.md#phantom-data
    _marker: PhantomData<T>,
}
```

Reply:
I copied the relevant code to a new project